### PR TITLE
feat: "kid" support in JWT header and multiple keys.

### PR DESCRIPTION
**Pull request by merlin-northern**
Ticket: MEN-6804
Changelog: title

### CODE REVIEW

**Comment by merlin-northern** [comment]
@mender-test-bot start pipeline please baby

**Comment by merlin-northern** [comment]
@mender-test-bot start pipeline please baby

**Comment by merlin-northern** [comment]
@mender-test-bot start pipeline please honey

File: authz/middleware.go [hunk 1]
```diff
@@ -54,7 +54,7 @@ func (mw *AuthzMiddleware) MiddlewareFunc(h rest.HandlerFunc) rest.HandlerFunc {
 		l := log.FromContext(r.Context())
 
 		//get token, no token header = http 401
-		tokstr, err := ExtractToken(r.Request)
+		tokstr, err := ExtractToken(r.Request) // here
```
Comment (kjaskiewiczz) [inline authz/middleware.go:57]: what's this comment about?

File: jwt/jwt_rsa.go [hunk 2]
```diff
@@ -15,38 +15,64 @@ package jwt
...
+	data, err := jt.SignedString(j.privKey[token.KeyId])
 	return data, err
 }
 
 func (j *JWTHandlerRS256) FromJWT(tokstr string) (*Token, error) {
 	jwttoken, err := jwt.ParseWithClaims(tokstr, &Claims{},
 		func(token *jwt.Token) (interface{}, error) {
+			keyId := common.KeyIdZero
+			if _, ok := token.Header["kid"]; ok {
+				keyId = int(token.Header["kid"].(float64)) // TODO: check types just in case
```
Comment (kjaskiewiczz) [inline jwt/jwt_rsa.go:67]: same here

**Review by alfrunes — COMMENTED** [review]
The title of the PR and the ticket description does not align.
Although I like the idea of adding support for JWK support in the backend, the ticket was actually about adding support for allowing tokens with no key signature. Probably the easiest way of achieving this is to set the `alg` header field [to `none`](https://datatracker.ietf.org/doc/html/rfc7519#section-6) if none of the private key configs are set.

File: user/useradm.go [hunk 3]
```diff
@@ -218,20 +230,22 @@ func (u *UserAdm) Login(ctx context.Context, email model.Email, pass string,
...
-		Issuer:    u.config.Issuer,
-		IssuedAt:  now,
-		NotBefore: now,
-		Tenant:    tenant,
-		Scope:     scope,
-		User:      true,
-	}}
+	ret := &jwt.Token{
+		KeyId: keyId,
+		Claims: jwt.Claims{ // here
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply user/useradm.go]: sorry, this is how I debug :)

File: server.go [hunk 4]
```diff
@@ -50,15 +55,42 @@ func RunServer(c config.Reader) error {
 
 	l := log.New(log.Ctx{})
 
-	authz := &SimpleAuthz{}
+	authorizer := &SimpleAuthz{} // here
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply server.go]: fixed. sorry for that, this is how I debug :)

File: jwt/jwt_rsa.go (hunk 2 above)
Comment (merlin-northern) [reply jwt/jwt_rsa.go]: fixed.

File: authz/middleware.go (hunk 1 above)
Comment (merlin-northern) [reply authz/middleware.go]: fixed. sorry for that, this is how I debug :)

**Comment by merlin-northern** [comment]
> The title of the PR and the ticket description does not align. Although I like the idea of adding support for JWK support in the backend, the ticket was actually about adding support for allowing tokens with no key signature. Probably the easiest way of achieving this is to set the `alg` header field [to `none`](https://datatracker.ietf.org/doc/html/rfc7519#section-6) if none of the private key configs are set.

thanks. if you like it, I consider it a good sign :)

**Review by kjaskiewiczz — APPROVED** [review]
looks good to me, but since it's not exactly what was requested in the ticket, I'd like to know @alfrunes (the author of the ticket) opinion before merging

**Comment by merlin-northern** [comment]
@mender-test-bot start pipeline please honey

File: rsa.in
Comment (kjaskiewiczz) [inline rsa.in:1]: I guess this file shouldn't be here?

File: ed25519.in
Comment (kjaskiewiczz) [inline ed25519.in:1]: this file also shouldn't be here

File: rsa.in
Comment (merlin-northern) [reply rsa.in:1]: correct. sorry, leftover input file that I used to generate the go code from :)

File: ed25519.in
Comment (merlin-northern) [reply ed25519.in:1]: correct. sorry, leftover input file that I used to generate the go code from :)

**Comment by merlin-northern** [comment]
@mender-test-bot start pipeline --fast please

**Pull request merged**